common/async.go

```diff
@@ -1,15 +1,59 @@
 package common
 
-import "sync"
-
-func Parallel(tasks ...func()) {
-	var wg sync.WaitGroup
-	wg.Add(len(tasks))
-	for _, task := range tasks {
-		go func(task func()) {
-			task()
-			wg.Done()
-		}(task)
+// val: the value returned after task execution.
+// err: the error returned during task completion.
+// abort: tells Parallel to return, whether or not all tasks have completed.
+type Task func(i int) (val interface{}, err error, abort bool)
+
+type TaskResult struct {
+	Value interface{}
+	Error error
+	Panic interface{}
+}
+
+type TaskResultCh <-chan TaskResult
+
+// Run tasks in parallel, with ability to abort early.
+// NOTE: Do not implement quit features here.  Instead, provide convenient
+// concurrent quit-like primitives, passed implicitly via Task closures. (e.g.
+// it's not Parallel's concern how you quit/abort your tasks).
+func Parallel(tasks ...Task) []TaskResultCh {
+	var taskResultChz = make([]TaskResultCh, len(tasks)) // To return.
+	var taskDoneCh = make(chan bool, len(tasks))         // A "wait group" channel, early abort if any true received.
+
+	// Start all tasks in parallel in separate goroutines.
+	// When the task is complete, it will appear in the
+	// respective taskResultCh (associated by task index).
+	for i, task := range tasks {
+		var taskResultCh = make(chan TaskResult, 1) // Capacity for 1 result.
+		taskResultChz[i] = taskResultCh
+		go func(i int, task Task, taskResultCh chan TaskResult) {
+			// Recovery
+			go func() {
+				if pnk := recover(); pnk != nil {
+					taskResultCh <- TaskResult{nil, nil, pnk}
+				}
```

Comment:
so, if we panic, we never write true to `taskDoneCh` and Parallel never finishes?

Reply:
oh we should write falase to taskDoneCh.

Reply:
that's strange, I wonder why the panic test didn't catch that.  I'll take a look.

Reply:
looks like the tests did catch it, I just never ran them properly :P
Please checkout the latest commit.